lib/forwarder/forwarder.go

```diff
@@ -0,0 +1,43 @@
+package forwarder
+
+import (
+	"context"
+	"io"
+	"sync"
+	"tcplb/lib/errors"
+)
+
+// MediocreForwarder is a implementation of the Forward operation.
+// This is a placeholder implementation that lacks robustness.
+type MediocreForwarder struct{}
+
+func (f MediocreForwarder) Forward(ctx context.Context, clientConn, upstreamConn DuplexConn) error {
+	// Caller is responsible for closing both DuplexConns, not us.
+	out := make(chan error, 2)
+	wg := sync.WaitGroup{}
+
+	copy := func(dst, src DuplexConn, out chan<- error) {
+		defer wg.Done()
+		// TODO FIXME add an idle timeout here that detects if neither
+		// of the two directions of copying have made any progress in
+		// some time window.
+		// TODO FIXME also honour cancellation by ctx
+		_, err := io.Copy(dst, src)
```

Comment:
As you hinted, this doesn't respect the context. Also, I wonder if it's worth continuing the other copy if one of the copies failed. Should we try and address those here?

Reply:
I think it is very important to address these, but i'd prefer to do it in a dedicated PR.

here's the rough approach i have in mind:

* use net.Conn SetDeadline to implement an idle timeout
* for context cancellation, we could react to that by calling SetDeadline to force an immediate timeout
* for copy failure, we could react to that by calling SetDeadline to force an immediate timeout
* when the copies halt after a timeout, we could check to see what the cause of the timeout was. if it was an error or a context cancellation/deadline/timeout, give up. if we timed out from the idle timeout but still appear to be making progress with one or both copies not at EOF, then reset the idle timeout, keep copying 

i probably need a few hours to work through all the details when i'm fresh